### PR TITLE
shadowcasting benchmark

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
SUMMARY: None

#### Purpose of change
Better shadowcasting benchamark.

#### Describe the solution
Added an actual 3D noise benchmark, a benchmark on an open plain, and an open plain with sparse obstacles.

#### Describe alternatives you've considered
There are innumerable other scenes worth adding.

#### Testing
It *is* a test.

### CODE REVIEW

File: tests/shadowcasting_test.cpp [hunk 1]
```diff
@@ -340,6 +340,77 @@ static void shadowcasting_float_quad(
     REQUIRE( passed );
 }
 
+static void do_3d_benchmark(
+    std::array<const float ( * )[MAPSIZE *SEEX][MAPSIZE *SEEY], OVERMAP_LAYERS> &transparency_caches,
+    const int iterations )
+{
+    float seen_squares[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{0}};
+    bool floor_cache[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{false}};
```
Comment (kevingranade) [inline tests/shadowcasting_test.cpp:348]:
```suggestion
    float seen_squares[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{{0}}};
    bool floor_cache[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{{false}}};
```

File: tests/shadowcasting_test.cpp [hunk 2]
```diff
@@ -340,6 +340,77 @@ static void shadowcasting_float_quad(
...
+        const long long diff =
+            std::chrono::duration_cast<std::chrono::microseconds>( end - start ).count();
+        printf( "cast_zlight() executed %d times in %lld microseconds.\n",
+                iterations, diff );
+    }
+}
+
+static void shadowcasting_3d_benchmark( const int iterations )
+{
+    float transparency_cache[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{0}};
```
Comment (kevingranade) [inline tests/shadowcasting_test.cpp:376]:
```suggestion
    float transparency_cache[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{{0}}};
```

File: tests/shadowcasting_test.cpp [hunk 3]
```diff
@@ -340,6 +340,77 @@ static void shadowcasting_float_quad(
     REQUIRE( passed );
 }
 
+static void do_3d_benchmark(
+    std::array<const float ( * )[MAPSIZE *SEEX][MAPSIZE *SEEY], OVERMAP_LAYERS> &transparency_caches,
+    const int iterations )
+{
+    float seen_squares[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {{{0}}};
```
Comment (jbytheway) [inline tests/shadowcasting_test.cpp:347]:
```suggestion
    float seen_squares[OVERMAP_LAYERS][MAPSIZE * SEEX][MAPSIZE * SEEY] = {};
```
If it looks cleaner to you, you can just initialize these arrays like this and it should be equivalent (and maybe won't cause compiler complaints?).

Comment (kevingranade) [reply tests/shadowcasting_test.cpp]: Looks good so far, I just hate the implicit-ness of it.

**Pull request merged**